### PR TITLE
Stretch cluster: Net split test case automation 

### PR DESCRIPTION
**Pull request by mashetty330**
This PR covers the netsplit test case scenarios for stretch cluster. Complete test plan can be found [here](https://url.corp.redhat.com/2424c21)

- [x] Run logwriter CephFS/logwriter RBD workload
- [ ]  Fetch ceph mon quorum and reset connection scores for each mons
- [x] Induce the network split between zones
- [x] Make sure there is no Data loss, Data corruption and Data unavailability
- [x] Make sure there is no issue from ceph side
- [ ] Make sure connection scores are clean post recovery
- [x] Make sure successful cluster recovery

### CODE REVIEW

File: tests/disaster-recovery/sc_arbiter/conftest.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+def pytest_collection_modifyitems(items):
+    """
+    A pytest hook to  filter out SC Arbiter tests
+
+    Args:
+        items: list of collected tests
+    """
+
+    if config.ENV_DATA.get("arbiter_deployment") is False:
+        for item in items.copy():
```
Comment (shivamdurgbuns) [inline tests/disaster-recovery/sc_arbiter/conftest.py:28]: I think `items.copy()` is not needed here if you are removing from items.  Simple `for item in items:` will do.

File: ocs_ci/ocs/constants.py [hunk 2]
```diff
@@ -2202,3 +2203,24 @@
     "rule3": UI_INPUT_RULES_GENERAL["rule2"],
     "rule4": UI_INPUT_RULES_GENERAL["rule3"],
 }
+
+# Stretch cluster
+STRETCH_CLUSTER_NAMESPACE = "sc-project"
+LOGREADER_JOB_PATH = os.path.join(TEMPLATE_STRETCH_CLUSTER_DIR, "logreader.cephfs.yaml")
```
Comment (PrasadDesala) [inline ocs_ci/ocs/constants.py:2209]: We already have the ceph logreader and writer yamls [here](https://github.com/red-hat-storage/ocs-ci/tree/master/ocs_ci/templates/workloads/logwriter) . Is this logreader something different than the existing one?

File: ocs_ci/ocs/resources/pod.py [hunk 3]
```diff
@@ -627,6 +627,17 @@ def get_node(self):
         else:
             return self.pod_data["spec"]["nodeName"]
 
+    def wait_for_pod_delete(self, resource_name=None, timeout=60):
+        """
+        Waits for the pod delete
+
```
Comment (PrasadDesala) [inline ocs_ci/ocs/resources/pod.py:633]: func args doc string missing

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+               Make sure the above pods run fine after the nodes are restarted
+            5) Delete the old logreader job and create new logreader job to verify the data corruption
+            6) Make sure there is no data loss
+            7) Do a complete cluster sanity and make sure there is no issue post recovery
+
+        """
+
+        # run cephfs workload for both logwriter and logreader
+        logwriter_workload, logreader_workload = setup_logwriter_cephfs_workload_factory
+        time.sleep(60)
```
Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:116]: as discussed, instead of adding sleep here add a logic to retry the failed command

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+            pytest.param(constants.NETSPLIT_ARBITER_DATA_1_AND_DATA_1_DATA_2, 15),
+        ],
+        ids=[
+            "Data-1-Data-2",
+            "Arbiter-Data-1",
+            "Arbiter-Data-1-and-Arbiter-Data-2",
+            "Arbiter-Data-1-and-Data-1-Data-2",
+        ],
+    )
+    def test_netsplit_cephfs(
```
Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:89]: Can we have both cephfs and rbd covered in the same test? Few of the steps in the test, setup, teardown are redundant. Combining these two tests can greatly reduce the efficiency and run time.

Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:89]: Can we have both cephfs and rbd covered in the same test? Few of the steps in the test, setup, teardown are redundant. Combining these two tests can greatly reduce the efficiency and run time.

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        for pod in pods_not_running:
+            node_obj = get_pod_node(pod)
+            nodes.stop_nodes(nodes=[node_obj], wait=False)
+            wait_for_nodes_status(
+                node_names=[node_obj.name], status=constants.NODE_NOT_READY
+            )
+            nodes.start_nodes(nodes=[node_obj])
+            wait_for_nodes_status(node_names=[node_obj.name])
+
+        wait_for_pods_to_be_in_statuses(
```
Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:240]: you have used this func in many places with a change in pod name. You can actually create a separate function for it 
def wait_for_pods_completion(pods, timeout=TEST_TIMEOUT):

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+                logger.info(f"Read success: {t_time}")
+            time_var = time_var + timedelta(minutes=1)
+        if pause_count > 5:
+            paused = True
+            break
+    return paused
+
+
+def check_for_write_pause(logwriter_pod, log_files, start_time, end_time):
+    paused = False
```
Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:49]: func doc string missing

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+            1) Run a logwriter RBD workload with RWO volumes
+            2) Check for write pause and fail when neccessary
+            3) Run logreader script inside the logwriter pods
+               and make sure no data corruption is seen
+            4) Make sure no data loss seen
+            5) Perfrom the complete cluste sanity and
+            make sure no issues post recovery
+
+        """
+        time.sleep(60)
```
Comment (PrasadDesala) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:353]: same as the comment for cephfs

File: tests/disaster-recovery/sc_arbiter/conftest.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    wait_for_pods_to_be_running,
+    get_deployments_having_label,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def pytest_collection_modifyitems(items):
+    """
+    A pytest hook to  filter out SC Arbiter tests
```
Comment (ebenahar) [inline tests/disaster-recovery/sc_arbiter/conftest.py:21]: if you are going to send another commit, please replace `SC` with `Stretched cluster`

File: tests/disaster-recovery/sc_arbiter/conftest.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+                label="app=logreader-cephfs", namespace=pvc.namespace
+            )
+        ]
+        wait_for_pods_to_be_running(
+            namespace=pvc.namespace, pod_names=logreader_job_pods
+        )
+
+        return logreader_job_obj
+
+    return factory
```
Comment (ebenahar) [inline tests/disaster-recovery/sc_arbiter/conftest.py:120]: These fixtures don't seem to be specific to arbiter. I suggest placing them in tests/conftest.py

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+    get_pods_having_label,
+    get_pod_logs,
+    get_pod_node,
+    get_ceph_tools_pod,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def check_for_read_pause(logreader_pods, start_time, end_time):
```
Comment (ebenahar) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:26]: how about creating a workload util file for logreader? The functionality could be used in other, non arbiter tests

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py (hunk 5 above)
Comment (mashetty330) [reply tests/disaster-recovery/sc_arbiter/test_netsplit.py:99]: @PrasadDesala I will keep this as such for now. I have to raise 1 more PR after this pr and https://github.com/red-hat-storage/ocs-ci/pull/7271 pr is merged, that time ill analyze and see if i can merge the two tests into two

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+
+        for pod_name in new_logreader_pods:
+            pod_logs = get_pod_logs(
+                pod_name=pod_name, namespace=constants.STRETCH_CLUSTER_NAMESPACE
+            )
+            assert "corrupt" not in pod_logs, "Data is corrupted!!"
+        logger.info("No data corruption is seen!")
+
+        # do cluster sanity post recovery
+        try:
```
Comment (parikshithb) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:306]: Can add this in teardown instead.

File: tests/disaster-recovery/sc_arbiter/test_netsplit.py [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+            output = logwriter_pod.exec_cmd_on_pod(
+                command=f"/opt/logreader.py -t 5 {list(log_file_map[logwriter_pod.name].keys())[0]} -d",
+                out_yaml_format=False,
+            )
+            assert "corrupt" not in output, "Data is corrupted!!"
+
+        logger.info("No data corruption is seen")
+
+        # do cluster sanity post recovery
+        try:
```
Comment (parikshithb) [inline tests/disaster-recovery/sc_arbiter/test_netsplit.py:422]: Same here

**Review by parikshithb — APPROVED** [review]
